### ISSUE TITLE
Rewrite it in Rust

### ISSUE THREAD

**Issue by WrinklyNinja** [WrinklyNinja/libloadorder#37]
Primarily as a learning exercise (more FFI, docs, libespm's ergonomics, a step up in complexity, etc.), but the library is a bit messy and could do with a refactor anyway. I'd like to avoid API breakage.

---

TODO:

- [x] Reimplement all get/set functionality
- [x] Reimplement thread-safe C FFI
- [x] Reimplement path caching, but only keep it if profiling shows it's beneficial
- [x] Reimplement desync detection
- [x] Reimplement on-disk load order validity detection
- [x] Convert docs

C API Breakage:

- String and array output needs to be explicitly freed in the Rust implementation, as their lifetimes are different. This could be avoided by having an internal static cache, but is that desirable?
- `LIBLO_ERROR_TIMESTAMP_READ_FAIL` is unused as read failures are reported as I/O errors. This change could potentially be avoided.
- If attempting to write an invalid filename, the Rust implementation errors, while the C++ implementation skips it and returns `LIBLO_WARN_BAD_FILENAME`. This could be avoided somehow, but it is due to invalid input, so an error is probably more correct.
- Memory allocation failure now causes a panic instead of a returned error code, so `LIBLO_ERROR_NO_MEM` is unused.
- The constness of some function arguments has changed, though this did not affect client code in the LOOT API.
- Attempting to deactivate an implicitly active plugin that isnot installed would not error in the C++ implementation, despite deactivating an implicitly active plugin and deactivating a plugin that is not installed both being error conditions. The Rust implementation fixes this bug.
- Attempting to set the active plugins giving an array including an implicitly active plugin that is not installed would not error in the C++ implementation, despite inclusion of a normal plugin that is not installed being an error condition. The Rust implementation fixes this bug.

If API breakage is embraced, then replacing the error codes with equivalents that map more cleanly to the Rust error types would be good.

**Comment by WrinklyNinja** [comment]
All the logic has now been implemented, the FFI is the only major work left to do.

**Comment by WrinklyNinja** [comment]
Well, damn, it turns out that you can't write a function that returns one of several types that implements a trait, eg.

```rust
fn load_order<T>(game_settings: GameSettings) -> T where T: WritableLoadOrder {
    match game_settings.load_order_method() {
        LoadOrderMethod::Asterisk => AsteriskBasedLoadOrder { game_settings, plugins: Vec::new() },
        LoadOrderMethod::Textfile => TextfileBasedLoadOrder { game_settings, plugins: Vec::new() },
        LoadOrderMethod::Timestamp => TimestampBasedLoadOrder { game_settings, plugins: Vec::new() },
    }
}
```

I tried with boxed traits too, but it looks like there just isn't a way to do it, which means I need to rework my implementation to not use traits and just match the load order method (like how the C++ code does).

**Comment by WrinklyNinja** [comment]
The Rust implementation currently has some differences:

- There's no path caching, only plugin caching. This should be profiled to see if the path caching is actually worth it.
- There's no desync detection, so `LIBLO_WARN_LO_MISMATCH` is unused.
- There's no validity check for the load order on disk, so `LIBLO_WARN_INVALID_LIST` is unused.
- `LIBLO_WARN_BAD_FILENAME` currently acts as a hard error.
- Strings returned from the FFI API need to be explicitly freed using helper functions.
- `lo_cleanup()` and `lo_is_compatible()` are now deprecated as they're unnecessary.
- Rust errors don't always map cleanly to the existing error codes.
- `LIBLO_ERROR_TIMESTAMP_READ_FAIL` is unused as it's treated as an I/O error.
- `LIBLO_ERROR_NO_MEM` is unused as memory allocation failure causes a panic.
- The constness of some FFI function arguments have changed as cbindgen is less opportunistic about adding const qualifiers, and Rust doesn't allow the objects pointed to by const pointers to be modified, unlike C++.

Given the intractability of some of the above points, it doesn't seem possible to make the API non-breaking, so I might as well tidy it up while I'm at it.

**Comment by WrinklyNinja** [comment]
The C++ headers generated need an extern "C" block around the inserted constant declarations (wrap them in #ifdef __cplusplus for C compatibility).

**Comment by WrinklyNinja** [comment]
The Rust FFI has been swapped into the LOOT API pretty smoothly (only a different include and a different library to link), but it looks like the lack of thread safety causes runtime issues, as when loading plugins concurrently the tests crash with `fatal runtime error: allocator memory exhausted`, but they succeed when loading plugins sequentially.

**Comment by WrinklyNinja** [comment]
Thread safety is in, but it seems like there's a terrible performance hit to the RwLock used, 5 concurrent checks if a plugin is active is taking 4 seconds in total in the LOOT API. This needs investigating...

**Comment by WrinklyNinja** [comment]
Some profiling results, with ~ 1000 plugins and 300 of them in the load order, using a release build of libloadorder. I'll update them as I test different games and caching combinations.

The test function used is available [here](https://gist.github.com/WrinklyNinja/f4b804bd223a845ef242373089517c1a) (with minor adaptations for the different load order types tested), and was built and run using `cargo test --release -- --nocapture --test-threads=1 profile_path_caching`.

All times are in seconds.

Asterisk

Caching | Cold Initial | Avg. Hot Initial | Avg. Repeated
----------|-------------|-----------------|----------------
None | 2.783 | 0.857 | 0.592
Plugin | 2.651 | 0.868 | 0.555
Path | | |
Plugin + Path | | |

Textfile

Caching | Cold Initial | Avg. Hot Initial | Avg. Repeated
----------|-------------|-----------------|----------------
None | 2.373 | 0.404 | 0.250
Plugin | 2.299 | 0.403 | 0.207
Path | | |
Plugin + Path | | |

Timestamp

Caching | Cold Initial | Avg. Hot Initial | Avg. Repeated
----------|-------------|-----------------|----------------
None | 0.563 |  0.137 | 0.163
Plugin | 0.533 | 0.136 | 0.139
Path | | |
Plugin + Path | | |

**Comment by WrinklyNinja** [comment]
I'd like to avoid caching entirely, as plugin timestamp locking is relatively common functionality that would prevent libloadorder detecting changes. One possible alternative would be to multithread reloading plugins, so try profiling that instead.

**Comment by WrinklyNinja** [comment]
It turns out that `LIBLO_WARN_INVALID_LIST` hasn't been used in the C++ API for a while now, so that validation is one less thing I need to reimplement.

**Comment by WrinklyNinja** [comment]
Updated profiling results as of 387429854f422dfcaa09f76adf10cb07db2ad68b:

Method | Cold Initial / s | Avg. Hot Initial / s | Avg. Repeated / s
----------|-------------|-----------------|----------------
Asterisk | 2.114 | 0.228 | 0.198
Textfile | 2.411 | 0.335 | 0.130
Timestamp | 0.492 | 0.117 | 0.119

However, something odd is going on, because the above suggests that getting the active state of 168 Fallout 4 plugins through the C API (load() + is_active()) would take ~ 35s, but profiling from a C++ test gives 3s. I've verified it's not just skipping everything by adding a `printf` line for each check, and they all happen within the 3s.

Loading the same number of plugins through LOOT takes 74s. It takes each plugin ~ 0.4s to check if its active, which ends up being the bottleneck, as each check locks the libloadorder game handle so they happen serially even though the plugins are loaded in parallel.

So in summary, if the timings from the Rust test are taken as 100% time, C++ says the times are 10% and LOOT says the times are 200% (though it has an abstraction layer, it's thin, so shouldn't be halving the speed).

**Comment by WrinklyNinja** [comment]
Updated profiling results as of 387429854f422dfcaa09f76adf10cb07db2ad68b:

Method | Cold / s | Avg. Hot / s
----------|---------|---------------
Asterisk | 1.47 | 0.141
Textfile | 2.54 | 0.340
Timestamp | 0.135 | 0.041

(Dropped the Avg. Repeated column now that there's no caching in libloadorder.)

The improvement is due to loading plugins in parallel when the plugins directory is scanned, and only loading them once, skipping the is_valid check and treating an error as an invalid plugin instead.

The Textfile method didn't benefit from this because its test case has all the plugins listed in loadorder.txt, so no plugins are loaded from the directory. Most of the different between Asterisk and Timestamp is also due to loading plugins from a text file, so that'll be my next focus.

**Comment by WrinklyNinja** [comment]
Closing this as the Rust rewrite has been merged into master as of ea1c9aa51f4e8be6d7283ecc1fa0274b1714a7eb.
